builtin/logical/pki/storage.go

```diff
@@ -0,0 +1,857 @@
+			existingIssuer.KeyID = result.ID
+			if err := writeIssuer(mkc.ctx, s, existingIssuer); err != nil {
+				return nil, false, err
+			}
+		}
+	}
+
+	// If there was no prior default value set and/or we had no known
+	// keys when we started, set this key as default.
+	keyDefaultSet, err := isDefaultKeySet(mkc.ctx, s)
+	if err != nil {
+		return nil, false, err
+	}
+	if len(knownKeys) == 0 || !keyDefaultSet {
+		if err = updateDefaultKeyId(mkc.ctx, s, result.ID); err != nil {
+			return nil, false, err
+		}
+	}
+
+	// All done; return our new key reference.
+	return &result, false, nil
+}
+
+func (i issuerEntry) GetCertificate() (*x509.Certificate, error) {
+	block, extra := pem.Decode([]byte(i.Certificate))
+	if block == nil {
+		return nil, errutil.InternalError{Err: fmt.Sprintf("unable to parse certificate from issuer: invalid PEM: %v", i.ID)}
+	}
+	if len(strings.TrimSpace(string(extra))) > 0 {
+		return nil, errutil.InternalError{Err: fmt.Sprintf("unable to parse certificate for issuer (%v): trailing PEM data: %v", i.ID, string(extra))}
+	}
+
+	return x509.ParseCertificate(block.Bytes)
+}
+
+func (i issuerEntry) EnsureUsage(usage issuerUsage) error {
+	// We want to spit out a nice error message about missing usages.
+	if i.Usage.HasUsage(usage) {
+		return nil
+	}
+
+	issuerRef := fmt.Sprintf("id:%v", i.ID)
+	if len(i.Name) > 0 {
+		issuerRef = fmt.Sprintf("%v / name:%v", issuerRef, i.Name)
+	}
+
+	// These usages differ at some point in time. We've gotta find the first
+	// usage that differs and return a logical-sounding error message around
+	// that difference.
+	for name, candidate := range namedIssuerUsages {
+		if usage.HasUsage(candidate) && !i.Usage.HasUsage(candidate) {
+			return fmt.Errorf("requested usage %v for issuer [%v] but only had usage %v", name, issuerRef, i.Usage.Names())
+		}
+	}
+
+	// Maybe we have an unnamed usage that's requested.
+	return fmt.Errorf("unknown delta between usages: %v -> %v / for issuer [%v]", usage.Names(), i.Usage.Names(), issuerRef)
+}
+
+func listIssuers(ctx context.Context, s logical.Storage) ([]issuerID, error) {
+	strList, err := s.List(ctx, issuerPrefix)
+	if err != nil {
+		return nil, err
+	}
+
+	issuerIds := make([]issuerID, 0, len(strList))
+	for _, entry := range strList {
+		issuerIds = append(issuerIds, issuerID(entry))
+	}
+
+	return issuerIds, nil
+}
+
+func resolveKeyReference(ctx context.Context, s logical.Storage, reference string) (keyID, error) {
+	if reference == defaultRef {
+		// Handle fetching the default key.
+		config, err := getKeysConfig(ctx, s)
+		if err != nil {
+			return keyID("config-error"), err
+		}
+		if len(config.DefaultKeyId) == 0 {
+			return KeyRefNotFound, fmt.Errorf("no default key currently configured")
+		}
+
+		return config.DefaultKeyId, nil
+	}
+
+	keys, err := listKeys(ctx, s)
+	if err != nil {
+		return keyID("list-error"), err
+	}
+
+	// Cheaper to list keys and check if an id is a match...
+	for _, keyId := range keys {
```

Comment:
A note from @sgmiller, he would prefer we use a Get operation to lookup a key id and only perform the list and iteration approach for name lookups (assuming the Get operation return nil,nil)

Reply:
Addressed in #15322